### PR TITLE
Run and save meta data saving.

### PR DESCRIPTION
**Pull request by pfafflabatuiuc**
Added the ability to save meta-data with the run_and_save_sweep() function. The function accepts any kwargs and will try to save it. The names of the files are the keys of the kwargs.

If a kwarg is a dictionary, it will try to save it as a json file. If it fails and detects that an ndarray is present as one of its values, it will transform it into a list. After trying, if it still fails, the dictionary will be pickled and saved.

Any other object passed will be pickled and saved. If an object is passed that cannot be pickled, the function will print a warning saying that the item was not able to be saved, but it will still run and save the sweep.

A few things I would like to have feedback:

- The whole try and save as a dictionary, convert and try again might be a little bit too complicated and could be simplified if either we just save everything as a pickle, or try to save a json once instead of converting and trying again.
- When trying to save to JSON but it fails, a file is still created with missing data. I could not properly figure out what objects get saved and which ones don't when this happens, so to not save extra dead files I manually delete them.
- General code style.

### CODE REVIEW

**Comment by wpfff** [comment]
let's look at the code together before merging.

File: labcore/ddh5.py [hunk 1]
```diff
@@ -73,20 +75,75 @@ def _check_none(line: Dict) -> bool:
...
     """
     data_dict = _create_datadict_structure(sweep)
 
     # Creates a file even when it fails.
     with DDH5Writer(data_dict, data_dir, name=name) as writer:
+
+        # Saving meta-data
+        dir = writer.filepath.removesuffix(writer.filename)
+        for key, value in extra_saving_items.items():
+            pickle_directory = dir + '\\' + key + '.pickle'
```
Comment (wpfff) [inline labcore/ddh5.py]: always use os.path.join for making paths.

File: labcore/ddh5.py [hunk 2]
```diff
@@ -73,20 +75,75 @@ def _check_none(line: Dict) -> bool:
...
+            if isinstance(value, dict):
+                dictionary_directory = dir + '\\' + key + '.json'
```
Comment (wpfff) [inline labcore/ddh5.py]: see above.

**Review by wpfff — CHANGES_REQUESTED** [review]
i think it would be cleaner to analyze the metadata for specific characteristics rather than doing everything with try/except blocks.
my approach would be:
- copy the dictionary
- walk through the data and convert things that need to be converted
- try to save
- only if that fails, pickle it

File: labcore/ddh5.py [hunk 3]
```diff
@@ -73,20 +75,75 @@ def _check_none(line: Dict) -> bool:
...
     """
     data_dict = _create_datadict_structure(sweep)
 
     # Creates a file even when it fails.
     with DDH5Writer(data_dict, data_dir, name=name) as writer:
+
+        # Saving meta-data
+        dir = writer.filepath.removesuffix(writer.filename)
+        for key, value in extra_saving_items.items():
+            pickle_directory = os.path.join(dir, key + '.pickle')
```
Comment (wpfff) [inline labcore/ddh5.py]: better naming here... this is not a directory, but a filepath, right?

File: labcore/ddh5.py [hunk 4]
```diff
@@ -73,20 +75,75 @@ def _check_none(line: Dict) -> bool:
...
+            if isinstance(value, dict):
+                dictionary_directory = os.path.join(dir, key + '.json')
+                try:
+                    _save_dictionary(value, dictionary_directory)
```
Comment (wpfff) [inline labcore/ddh5.py]: see above.

File: labcore/ddh5.py [hunk 5]
```diff
@@ -73,20 +75,75 @@ def _check_none(line: Dict) -> bool:
...
+                except TypeError as error:
+                    # Delete the file created by _save_dictionary. This file does not contain the complete dictionary.
+                    if os.path.isfile(dictionary_directory):
+                        os.remove(dictionary_directory)
+
+                    converted = False # Flag to see if there has been a converted ndarray.
+                    for k, v in value.items():
```
Comment (wpfff) [inline labcore/ddh5.py:123]: why are you converting and then saving to pickle?
i thought the whole point of converting to list was that we can save as json?

**Review by wpfff — CHANGES_REQUESTED** [review]

File: labcore/ddh5.py (hunk 5 above)
Comment (pfafflabatuiuc) [reply labcore/ddh5.py:123]: This was a mistake. It is corrected now. Will write tests for it tomorrow.

**Pull request merged**